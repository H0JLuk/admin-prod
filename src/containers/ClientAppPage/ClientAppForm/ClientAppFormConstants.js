import { MECHANICS_CHECKBOXES, APP_LOGIN_TYPES, APP_MECHANICS, MECHANICS_ERROR } from '../../../constants/clientAppsConstants';
import { showCount } from '../../../constants/common';

const defaultImg = 'default_main_illustration';
const february23 = 'February23_main_illustration';
const march8 = 'March8_main_illustration';

export const CREATE_APP_TITLE = 'Новое приложение';
export const CANCEL_BUTTON_TITLE = 'Отменить';
export const ADD_BUTTON_TITLE = 'Добавить';
export const PROPERTIES_TITLE = 'Свойства';
export const EDIT_BUTTON_LABEL = 'Сохранить';
export const DESIGN_RADIO_TITLE = 'Оформление';
export const PROPERTIES_RADIO_TITLE = 'Свойства';
export const SUCCESS_MESSAGE = 'Операция завершена';
export const ERROR_MESSAGE = 'Ошибка';
export const SUCCESS_DESIGN_UPDATE_DESCRIPTION = 'Оформление для клиентского приложения успешно обновлено';
export const SUCCESS_PROPERTIES_UPDATE_DESCRIPTION = 'Настройки клиентского приложения успешно обновлены';
export const SUCCESS_PROPERTIES_CREATE_DESCRIPTION = 'Клиентское приложение успешно создано';
export const BACKEND_ERROR_ALREADY_EXIST_ENDING = 'already exists';

export const EDIT_MODE = {
    DESIGN: 'Оформление',
    PROPERTIES: 'Свойства',
};

export const keysToString = ['mechanics', 'login_types', 'where_to_use', 'design_elements'];

const numberTransform = value => value ? Number(value) : '';

export const TextKeysWithDefaultValues = ['home_page_header_present', 'home_page_header_wow'];

const checkBoxValidator = (_, value) => {
    if (
        value.includes(APP_MECHANICS.PRESENTS.value) ||
        value.includes(APP_MECHANICS.ECOSYSTEM.value) ||
        value.includes(APP_MECHANICS.BUNDLE.value)
    ) {
        return Promise.resolve();
    }
    return Promise.reject(MECHANICS_ERROR);
};

const DEFAULT_BANNER = {
    vitrina_theme: defaultImg,
    gradient: '228.35deg, #eaef00 7.94%, #cbea0a 15.79%, #9cdd16 21.25%, #5dca23 28.58%, #00b944 38.54%, #00a3a7 57.42%, #009bc5 64.31%, #0096d3 67.8%, #018fdb 72.33%, #0489d6 76.41%, #047fcf 82.5%, #0073c8 87.4%, #0051b7 99.87%',
    design_elements: [],
};

export const BANNER_KEYS = Object.keys(DEFAULT_BANNER);

export const banners = [
    { ...DEFAULT_BANNER },
    {
        vitrina_theme: february23,
        gradient: '262.6deg, #9cdd16 -9.99%, #84e73b 7.47%, #7dea44 13.24%, #18cd6d 32.14%, #06ac9a 43.29%, #00a3a7 51.61%, #00a0b2 57.73%, #0098b6 67.72%, #0073c8 90.72%',
        design_elements: [],
    },
    {
        vitrina_theme: march8,
        gradient: '269.17deg, #ffeb37 3.13%, #ff8c39 50.88%, #f01d71 99.82%',
        design_elements: [],
    },
];

export const DEFAULT_DESIGN_SETTINGS = {
    ...DEFAULT_BANNER,
    home_page_header: 'Сбер изменился, чтобы стать еще ближе к вам',
};

export const designKeysForCheck = [...Object.keys(DEFAULT_DESIGN_SETTINGS), ...TextKeysWithDefaultValues];

const appCodeRegex = /^[a-z0-9-_]+$/gmi;

export const RULES = {
    STANDARD: [],
    STANDARD_NUMBER: [
        {
            type: 'number',
            transform: numberTransform,
            validateTrigger: 'onSubmit',
            message: 'Значение может быть только числовым',
        },
    ],
    TWO_DIGITS_NUMBER: [
        {
            type: 'number',
            transform: numberTransform,
            validateTrigger: 'onSubmit',
            message: 'Значение может быть только числовым',
        },
    ],
    CHECKBOX_REQUIRED : [
        { type: 'array', required: true, message: 'Это обязательное поле', validateTrigger: 'onSubmit' },
    ],
    CHECKBOX: [
        { type: 'array', required: true, message: 'Это обязательное поле', validateTrigger: 'onSubmit' },
        { type: 'array', validator: checkBoxValidator, validateTrigger: 'onSubmit' },
    ],
    REQUIRED: [{ required: true, message: 'Это обязательное поле', validateTrigger: 'onSubmit' }],
    REQUIRED_ENGLISH_AND_NUMBER: [
        { required: true, message: 'Это обязательное поле' },
        { pattern: appCodeRegex, message: 'Можно использовать только латиницу', validateTrigger: 'onSubmit' },
    ],
};

export const FORM_TYPES = {
    INPUT: 'INPUT',
    CHECKBOX_GROUP: 'CHECKBOX_GROUP',
    TEXT_BLOCK: 'TEXT_BLOCK',
    MAIN_INFO_INPUT: 'MAIN_INFO_INPUT',
    BANNER: 'BANNER',
    SELECT: 'SELECT',
};

export const SETTINGS_TYPES = {
    CREATE: 'create',
    EDIT: 'edit',
};

export const FORM_MODES = {
    NEW: 'new',
    EDIT: 'edit',
};

export const mainInfoElements = [
    [
        {
            label: 'Название',
            type: FORM_TYPES.MAIN_INFO_INPUT,
            span: 16,
            rules: RULES.REQUIRED,
            name: 'name',
            placeholder: 'Имя',
        },
        {
            label: 'Код',
            type: FORM_TYPES.MAIN_INFO_INPUT,
            span: 8,
            rules: RULES.REQUIRED_ENGLISH_AND_NUMBER,
            name: 'code',
            placeholder: 'Код',
        },
    ],
    [
        {
            label: 'Отображаемое имя',
            type: FORM_TYPES.MAIN_INFO_INPUT,
            span: 16,
            rules: RULES.REQUIRED,
            name: 'displayName',
            placeholder: 'Отображаемое имя',
        },
    ]
];

export const formElements = [
    [
        {
            label: 'Токен Яндекс Метрики',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: RULES.STANDARD_NUMBER,
            name: 'ym_token',
            placeholder: 'Токен',
            maxLength: 12,
        },
        {
            label: 'Сессия сотрудника (в секундах)',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: RULES.STANDARD_NUMBER,
            name: 'token_lifetime',
            placeholder: '1800 секунд по умолчанию',
            maxLength: 12,
        },
        {
            label: 'Сессия клиента (в секундах)',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: RULES.STANDARD_NUMBER,
            name: 'inactivity_time',
            placeholder: '15 секунд по умолчанию',
            maxLength: 12,
        },
    ],

    [
        {
            label: 'Показ предложения в "карусели" (в секундах)',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: RULES.STANDARD_NUMBER,
            name: 'promo_show_time',
            placeholder: '20 секунд по умолчанию',
            maxLength: 12,
        },
        {
            label: 'Максимальное число попыток входа',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: RULES.TWO_DIGITS_NUMBER,
            name: 'max_password_attempts',
            placeholder: '3 по умолчанию',
            maxLength: 2,
        },
        {
            label: 'Временная блокировка пользователя (в секундах)',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: RULES.STANDARD_NUMBER,
            name: 'tmp_block_time',
            placeholder: '1800 секунд по умолчанию',
            maxLength: 12,
        },
    ],

    [
        {
            label: 'Максимальное число подарков',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: RULES.TWO_DIGITS_NUMBER,
            name: 'max_presents_number',
            placeholder: '3 по умолчанию',
            maxLength: 2,
        },
        {
            label: 'Путь к инструкции по установке',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: RULES.STANDARD,
            name: 'installation_url',
            placeholder: 'url',
        },
        {
            label: 'Путь к инструкции по использованию',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: RULES.STANDARD,
            name: 'usage_url',
            placeholder: 'url',
        },
    ],

    [
        {
            label: 'Политика конфиденциальности',
            type: FORM_TYPES.TEXT_BLOCK,
            span: 13,
            rows: 3,
            rules: RULES.STANDARD,
            name: 'privacyPolicy',
            placeholder: 'Описание ДЗО',
        },
        {
            label: 'Способ авторизации в витрине',
            type: FORM_TYPES.CHECKBOX_GROUP,
            options: APP_LOGIN_TYPES,
            rules: RULES.CHECKBOX_REQUIRED,
            columnMode: true,
            span: 11,
            name: 'login_types',
        },
    ],
    [
        {
            label: 'Механика',
            type: FORM_TYPES.CHECKBOX_GROUP,
            span: 24,
            rules: RULES.CHECKBOX,
            options: MECHANICS_CHECKBOXES,
            name: 'mechanics',
        },
    ]
];

export const designElements = [
    [
        {
            label: 'Текст для главной(Продукты)',
            type: FORM_TYPES.TEXT_BLOCK,
            span: 12,
            rows: 3,
            maxLength: 70,
            showCount,
            name: 'home_page_header',
            placeholder: 'Текст для главной(Продукты)',
        },
        {
            label: 'Текст для главной(Подарки)',
            type: FORM_TYPES.TEXT_BLOCK,
            span: 12,
            rows: 3,
            maxLength: 70,
            showCount,
            name: 'home_page_header_present',
            placeholder: 'Текст для главной(Подарки)',
        },
    ],
    [
        {
            label: 'Текст для главной(WOW)',
            type: FORM_TYPES.TEXT_BLOCK,
            span: 12,
            rows: 3,
            maxLength: 70,
            showCount,
            name: 'home_page_header_wow',
            placeholder: 'Текст для главной(WOW)',
        },
        {
            label: 'Тема',
            type: FORM_TYPES.BANNER,
            span: 12,
            name: 'home_page_theme',
        },
    ],
];