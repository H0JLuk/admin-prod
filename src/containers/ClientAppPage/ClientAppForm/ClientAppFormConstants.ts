import {
    APP_MECHANIC,
    GAME_MECHANIC_OPTIONS,
    APP_MECHANIC_OPTIONS,
    MECHANICS_ERROR,
    NOTIFICATION_TYPES_OPTIONS,
} from '@constants/clientAppsConstants';
import { LOGIN_TYPE_OPTIONS } from '@constants/loginTypes';
import { showCount } from '@constants/common';
import { getPatternAndMessage } from '@utils/validators';
import { ValidatorRule, Rule } from 'rc-field-form/lib/interface';
import { FormItemProps, InputProps, SelectProps } from 'antd';
import { CheckboxGroupProps } from 'antd/lib/checkbox';
import { TextAreaProps } from 'antd/lib/input';
import { BannerProps, IBanner } from './MainPageDesign/Banner';

const defaultImg = 'default_main_illustration';
const february23 = 'February23_main_illustration';
const march8 = 'March8_main_illustration';

export const CREATE_APP_TITLE = 'Новое приложение';
export const CANCEL_BUTTON_TITLE = 'Отменить';
export const ADD_BUTTON_TITLE = 'Добавить';
export const PROPERTIES_TITLE = 'Свойства';
export const EDIT_BUTTON_LABEL = 'Сохранить';
export const DESIGN_RADIO_TITLE = 'Оформление';
export const PROPERTIES_RADIO_TITLE = 'Свойства';
export const SUCCESS_MESSAGE = 'Операция завершена';
export const ERROR_MESSAGE = 'Ошибка';
export const SUCCESS_DESIGN_UPDATE_DESCRIPTION = 'Оформление для клиентского приложения успешно обновлено';
export const SUCCESS_PROPERTIES_UPDATE_DESCRIPTION = 'Настройки клиентского приложения успешно обновлены';
export const SUCCESS_PROPERTIES_CREATE_DESCRIPTION = 'Клиентское приложение успешно создано';
export const BACKEND_ERROR_ALREADY_EXIST_ENDING = 'already exists';
export const BUSINESS_ROLE_FOR_APPLICATION = 'Бизнес-роль, для которой доступно приложение';
export const BUSINESS_ROLE_FOR_APP_PLACEHOLDER = 'Выберите роль/роли';

export const EDIT_MODE = {
    DESIGN: 'Оформление',
    PROPERTIES: 'Свойства',
};

export const keysToString = [
    'mechanics',
    'login_types',
    'where_to_use',
    'design_elements',
    'notification_types',
    'game_mechanics',
];

const numberTransform = (value: string) => value ? Number(value) : '';

export const TextKeysWithDefaultValues = ['home_page_header_present', 'home_page_header_wow'];

const checkBoxValidator: ValidatorRule['validator'] = (_, value) => {
    if (
        value.includes(APP_MECHANIC.PRESENTS) ||
        value.includes(APP_MECHANIC.ECOSYSTEM) ||
        value.includes(APP_MECHANIC.BUNDLE)
    ) {
        return Promise.resolve();
    }
    return Promise.reject(MECHANICS_ERROR);
};

const DEFAULT_BANNER = {
    vitrina_theme: defaultImg,
    gradient: '228.35deg, #eaef00 7.94%, #cbea0a 15.79%, #9cdd16 21.25%, #5dca23 28.58%, #00b944 38.54%, #00a3a7 57.42%, #009bc5 64.31%, #0096d3 67.8%, #018fdb 72.33%, #0489d6 76.41%, #047fcf 82.5%, #0073c8 87.4%, #0051b7 99.87%',
    design_elements: [],
};

export const BANNER_KEYS = Object.keys(DEFAULT_BANNER) as (keyof IBanner)[];

export const banners: IBanner[] = [
    { ...DEFAULT_BANNER },
    {
        vitrina_theme: february23,
        gradient: '262.6deg, #9cdd16 -9.99%, #84e73b 7.47%, #7dea44 13.24%, #18cd6d 32.14%, #06ac9a 43.29%, #00a3a7 51.61%, #00a0b2 57.73%, #0098b6 67.72%, #0073c8 90.72%',
        design_elements: [],
    },
    {
        vitrina_theme: march8,
        gradient: '269.17deg, #ffeb37 3.13%, #ff8c39 50.88%, #f01d71 99.82%',
        design_elements: [],
    },
];

type IDefaultDesignSettings = Record<string, string | string[]>;

export const DEFAULT_DESIGN_SETTINGS: IDefaultDesignSettings = {
    ...DEFAULT_BANNER,
    home_page_header: 'Сбер изменился, чтобы стать еще ближе к вам',
};

export const designKeysForCheck = [...Object.keys(DEFAULT_DESIGN_SETTINGS), ...TextKeysWithDefaultValues];

const REQUIRED_VALIDATE_MESSAGE = 'Это обязательное поле';
const NUMBER_VALIDATE_MESSAGE = 'Значение может быть только числовым';

export const RULES = {
    REQUIRED: {
        required: true,
        message: REQUIRED_VALIDATE_MESSAGE,
        validateTrigger: 'onSubmit',
    } as Rule,
    NUMBER: {
        type: 'number',
        message: NUMBER_VALIDATE_MESSAGE,
        transform: numberTransform,
        validateTrigger: 'onSubmit',
    } as Rule,
    get CHECKBOX_GROUP() {
        return {
            ...this.REQUIRED,
            type: 'array',
        } as Rule;
    },
};

export enum FORM_TYPES {
    INPUT = 'INPUT',
    CHECKBOX_GROUP = 'CHECKBOX_GROUP',
    TEXT_AREA = 'TEXT_AREA',
    BANNER = 'BANNER',
    SELECT = 'SELECT',
}

export enum SETTINGS_TYPES {
    CREATE = 'create',
    EDIT = 'edit',
}

export enum FORM_MODES {
    NEW ='new',
    EDIT = 'edit',
}

export type CommonFormConstructorItem = {
    label: string;
    span: number | string;
    rules?: FormItemProps['rules'];
    name: string;
    canEdit?: boolean;
};

export type FormConstructorInput = Omit<InputProps, 'type'> & {
    type: FORM_TYPES.INPUT;
};

export type FormConstructorCheckboxGroup = CheckboxGroupProps & {
    type: FORM_TYPES.CHECKBOX_GROUP;
    columnMode?: boolean;
};

export type FormConstructorTextArea = TextAreaProps & {
    type: FORM_TYPES.TEXT_AREA;
};

export type FormConstructorBanner = BannerProps & {
    type: FORM_TYPES.BANNER;
};

export type FormConstructorSelect = SelectProps<any> & {
    type: FORM_TYPES.SELECT;
};

export type FormConstructorFormItem =
    FormConstructorInput |
    FormConstructorCheckboxGroup |
    FormConstructorTextArea |
    FormConstructorBanner |
    FormConstructorSelect;

export type FormConstructorItem = FormConstructorFormItem & CommonFormConstructorItem;

export const PRIVACY_POLICY = {
    TITLE: 'Согласие на обработку персональных данных',
    VERSION: 'Версия',
    START_DATE: 'Начало действия',
    DETAILS: 'Подробнее',
    EMPTY_CONSENTS: 'Согласия для витрины отсутствуют',
    TO_CONSENTS_LIST: 'Перейти к списку согласий',
};

export const mainInfoElements: FormConstructorItem[][] = [
    [
        {
            label: 'Название',
            type: FORM_TYPES.INPUT,
            span: 16,
            rules: [
                RULES.REQUIRED,
                {
                    ...getPatternAndMessage('clientApp', 'name'),
                    validateTrigger: 'onSubmit',
                },
            ],
            name: 'name',
            placeholder: 'Имя',
        },
        {
            label: 'Код',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: [
                RULES.REQUIRED,
                {
                    ...getPatternAndMessage('clientApp', 'code'),
                    validateTrigger: 'onSubmit',
                },
            ],
            name: 'code',
            placeholder: 'Код',
        },
    ],
    [
        {
            label: 'Отображаемое имя',
            type: FORM_TYPES.INPUT,
            span: 16,
            canEdit: true,
            rules: [
                RULES.REQUIRED,
                {
                    ...getPatternAndMessage('clientApp', 'name'),
                    validateTrigger: 'onSubmit',
                },
            ],
            name: 'displayName',
            placeholder: 'Отображаемое имя',
        },
    ],
];

export const formElements: FormConstructorItem[][] = [
    [
        {
            label: 'Сессия сотрудника (в секундах)',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: [
                RULES.NUMBER,
            ],
            name: 'token_lifetime',
            placeholder: '1800 секунд по умолчанию',
            maxLength: 12,
        },
        {
            label: 'Сессия клиента (в секундах)',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: [
                RULES.NUMBER,
            ],
            name: 'inactivity_time',
            placeholder: '15 секунд по умолчанию',
            maxLength: 12,
        },
        {
            label: 'Временная блокировка пользователя (в секундах)',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: [
                RULES.NUMBER,
            ],
            name: 'tmp_block_time',
            placeholder: '1800 секунд по умолчанию',
            maxLength: 12,
        },
    ],

    [
        {
            label: 'Сессия referalLink (в секундах)',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: [
                RULES.NUMBER,
            ],
            name: 'referal_lifetime',
            placeholder: '30 секунд по умолчанию',
            maxLength: 12,
        },
        {
            label: 'Максимальное число попыток входа',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: [
                RULES.NUMBER,
            ],
            name: 'max_password_attempts',
            placeholder: '3 по умолчанию',
            maxLength: 2,
        },
        {
            label: 'Максимальное число подарков',
            type: FORM_TYPES.INPUT,
            span: 8,
            rules: [
                RULES.NUMBER,
            ],
            name: 'max_presents_number',
            placeholder: '3 по умолчанию',
            maxLength: 2,
        },
    ],

    [
        {
            label: 'Механика',
            type: FORM_TYPES.CHECKBOX_GROUP,
            span: 12,
            rules: [
                RULES.CHECKBOX_GROUP,
                {
                    validator: checkBoxValidator,
                    validateTrigger: 'onSubmit',
                },
            ],
            options: APP_MECHANIC_OPTIONS,
            name: 'mechanics',
        },
        {
            label: 'Игровая механика',
            type: FORM_TYPES.CHECKBOX_GROUP,
            span: 12,
            options: GAME_MECHANIC_OPTIONS,
            name: 'game_mechanics',
        }
    ],
    [
        {
            label: 'Способ авторизации в витрине',
            type: FORM_TYPES.CHECKBOX_GROUP,
            options: LOGIN_TYPE_OPTIONS,
            rules: [
                RULES.CHECKBOX_GROUP,
            ],
            columnMode: true,
            span: 12,
            name: 'login_types',
        },
        {
            label: 'Способ отправки сообщений клиенту',
            type: FORM_TYPES.CHECKBOX_GROUP,
            span: 12,
            options: NOTIFICATION_TYPES_OPTIONS,
            name: 'notification_types',
        },
    ],
];

export const designElements: FormConstructorItem[][] = [
    [
        {
            label: 'Текст для главной(Продукты)',
            type: FORM_TYPES.TEXT_AREA,
            span: 12,
            rows: 3,
            maxLength: 70,
            showCount,
            name: 'home_page_header',
            placeholder: 'Текст для главной(Продукты)',
        },
        {
            label: 'Текст для главной(Подарки)',
            type: FORM_TYPES.TEXT_AREA,
            span: 12,
            rows: 3,
            maxLength: 70,
            showCount,
            name: 'home_page_header_present',
            placeholder: 'Текст для главной(Подарки)',
        },
    ],
    [
        {
            label: 'Текст для главной(WOW)',
            type: FORM_TYPES.TEXT_AREA,
            span: 12,
            rows: 3,
            maxLength: 70,
            showCount,
            name: 'home_page_header_wow',
            placeholder: 'Текст для главной(WOW)',
        },
        {
            label: 'Тема',
            type: FORM_TYPES.BANNER,
            span: 12,
            name: 'home_page_theme',
        },
    ],
];
