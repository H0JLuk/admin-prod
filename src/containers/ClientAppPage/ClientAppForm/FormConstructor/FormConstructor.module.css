/* colFlex */
.colFlex :global(.ant-checkbox-group) {
    display: flex;
    flex-wrap: wrap;
}
/* /colFlex */

.infoText {
    font-size: 14px;
    color: #4e4e4e;
}

.checkboxColumn {
    flex-direction: column;
}

.labelBold :global(.ant-form-item-control-input) {
    min-height: unset;
}

.labelBold > div > label {
    font-weight: bold;
}

/* hide required red star */
.labelBold > div > label::before {
    content: none !important;
}

.propertiesRow:not(:nth-last-child(3)) {
    align-items: flex-end;
}

.propertiesRow:nth-child(4) {
    align-items: flex-start;
}

.propertiesRow:nth-child(4) :global(.ant-row) {
    margin-bottom: 0 !important;
}

.propertiesRow:not(:last-child) {
    margin-bottom: 5px;
}
