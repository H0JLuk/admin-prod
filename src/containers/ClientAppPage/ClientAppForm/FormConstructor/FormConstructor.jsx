import React from 'react';
import cn from 'classnames';
import { Checkbox, Col, Form, Input, Row } from 'antd';
import { FORM_TYPES } from '../ClientAppFormConstants';
import Banner from '../MainPageDesign/Banner/Banner';

import styles from './FormConstructor.module.css';

function FormInputByType({
    isEdit,
    type,
    columnMode,
    ...restProps
}) {

    switch (type) {
        case FORM_TYPES.BANNER:
            return <Banner { ...restProps } />;
        case FORM_TYPES.CHECKBOX_GROUP:
            return (
                <Checkbox.Group
                    className={ cn({ [styles.checkboxColumn]: columnMode }) }
                    { ...restProps }
                />
            );
        case FORM_TYPES.TEXT_BLOCK:
            return <Input.TextArea { ...restProps } />; /* TODO поменять во всем проекте на showCount */
        default:
            if (!isEdit) {
                return <Input allowClear { ...restProps } />;
            }
            return <div className={ styles.infoText }>{ restProps.value }</div>;
    }
}

export default function AppFormConstructor({ row, isEdit }) {
    return (
        <Row className={ styles.propertiesRow } gutter={ [24] }>
            { row.map(({ label, span, rules, name, ...restProps }) => (
                <Col className={ styles.colFlex } key={ label } span={ span }>
                    <Form.Item
                        rules={ rules }
                        name={ name }
                        className={ cn({ [styles.labelBold]: isEdit }) }
                        label={ label }
                        validateFirst
                    >
                        <FormInputByType isEdit={ isEdit } { ...restProps } />
                    </Form.Item>
                </Col>
            )) }
        </Row>
    );
}