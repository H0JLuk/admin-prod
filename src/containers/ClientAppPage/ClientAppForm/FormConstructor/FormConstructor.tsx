import React from 'react';
import cn from 'classnames';
import { Checkbox, Col, Form, Input, Row, Select } from 'antd';
import { FormConstructorItem, FormConstructorFormItem, FORM_TYPES } from '../ClientAppFormConstants';
import Banner from '../MainPageDesign/Banner';

import styles from './FormConstructor.module.css';

type FormInputByTypeProps = FormConstructorFormItem & {
    isEdit?: boolean;
    canEdit?: boolean;
    disabledFields?: Record<string, string[]>;
};

type AppFormConstructorProps = {
    isEdit?: boolean;
    row: FormConstructorItem[];
    disabledFields?: Record<string, string[]>;
    isCreate?: boolean;
};

const FormInputByType: React.FC<FormInputByTypeProps> = ({ isEdit, disabledFields = {}, ...rest }) => {
    switch (rest.type) {
        case FORM_TYPES.BANNER: {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const { type, ...restProps } = rest;
            return <Banner {...restProps} />;
        }
        case FORM_TYPES.CHECKBOX_GROUP: {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const { type, columnMode, ...restProps } = rest;
            const hasFieldsToDisable = disabledFields[(restProps as any).id];
            if (hasFieldsToDisable) {
                restProps.options = (restProps.options || []).map(
                    op => typeof op === 'object' ? ({ ...op, disabled: hasFieldsToDisable.includes(op.value as string) }) : op,
                );
            }
            return (
                <Checkbox.Group
                    className={cn({ [styles.checkboxColumn]: columnMode })}
                    {...restProps}
                />
            );
        }
        case FORM_TYPES.TEXT_AREA: {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const { type, ...restProps } = rest;
            return <Input.TextArea {...restProps} />;
        }
        case FORM_TYPES.INPUT: {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const { type, canEdit, ...restProps } = rest;
            if (!isEdit || canEdit) {
                return <Input allowClear {...restProps} />;
            }
            return <div className={styles.infoText}>{restProps.value}</div>;
        }
        case FORM_TYPES.SELECT: {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const { type, ...restProps } = rest;
            return <Select {...restProps} />;
        }
    }
};

const AppFormConstructor: React.FC<AppFormConstructorProps> = ({
    row,
    isEdit,
    isCreate = false,
    disabledFields,
}) => (
    <Row className={styles.propertiesRow} gutter={24}>
        {row.map(({ label, span, rules, name, normalize, hideWhenCreate, ...restProps }) => !(isCreate && hideWhenCreate) && (
            <Col className={styles.colFlex} key={label} span={span}>
                <Form.Item
                    rules={rules}
                    name={name}
                    className={cn({ [styles.labelBold]: isEdit })}
                    label={label}
                    normalize={normalize}
                    validateTrigger="onSubmit"
                    validateFirst
                >
                    <FormInputByType
                        isEdit={isEdit}
                        {...restProps}
                        disabledFields={disabledFields}
                    />
                </Form.Item>
            </Col>
        ))}
    </Row>
);

export default AppFormConstructor;
