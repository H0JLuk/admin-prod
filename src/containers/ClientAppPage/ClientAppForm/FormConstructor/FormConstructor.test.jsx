import React from 'react';
import { shallow } from 'enzyme';
import FormConstructor from './FormConstructor';
import { FORM_TYPES } from '../ClientAppFormConstants';

describe('<FormConstructor /> test', () => {
    const props = {
        isEdit: false,
        row: [
            {
                label: 'label',
                span: 'span',
                rules: 'rules',
                name: 'name',
                value: {
                    vitrina_theme: 'vitrina_theme',
                    gradient: 'gradient',
                },
            },
        ],
    };

    it('should be mount snap', () => {
        const FormConstructorComponent = shallow(
            <FormConstructor { ...props } />
        );

        expect(FormConstructorComponent.debug()).toMatchSnapshot();
    });

    it('should render bannerType', () => {
        const newProps = {
            ...props,
            row: [{ ...props.row[0], type: FORM_TYPES.BANNER }],
        };
        const FormConstructorComponent = shallow(
            <FormConstructor { ...newProps } />
        );

        expect(FormConstructorComponent.find('FormInputByType').html()).toMatch('imageBlock');
        expect(FormConstructorComponent.find('FormInputByType').html()).toMatchSnapshot();
    });

    it('should render checkBoxes', () => {
        const newProps = {
            ...props,
            row: [{ ...props.row[0], type: FORM_TYPES.CHECKBOX_GROUP }],
        };
        const FormConstructorComponent = shallow(
            <FormConstructor { ...newProps } />
        );

        expect(FormConstructorComponent.find('FormInputByType').html()).toMatch('checkbox');
        expect(FormConstructorComponent.find('FormInputByType').html()).toMatchSnapshot();
    });

    it('should render textBlock', () => {
        const newProps = {
            ...props,
            row: [{ ...props.row[0], type: FORM_TYPES.TEXT_BLOCK }],
        };
        const FormConstructorComponent = shallow(
            <FormConstructor { ...newProps } />
        );

        expect(FormConstructorComponent.find('FormInputByType').html()).toMatch('textarea');
        expect(FormConstructorComponent.find('FormInputByType').html()).toMatchSnapshot();
    });

    it('should render input in default', () => {
        const FormConstructorComponent = shallow(
            <FormConstructor { ...props } />
        );

        expect(FormConstructorComponent.find('FormInputByType').html()).toMatch('input');
        expect(FormConstructorComponent.find('FormInputByType').html()).toMatchSnapshot();
    });

    it('should render div in default with editMode', () => {
        const newProps = {
            ...props,
            isEdit: true,
            row: [{ ...props.row[0], value: 'test' }],
        };
        const FormConstructorComponent = shallow(
            <FormConstructor { ...newProps } />
        );

        expect(FormConstructorComponent.find('FormInputByType').html()).toMatch('infoText');
        expect(FormConstructorComponent.find('FormInputByType').html()).toMatchSnapshot();
    });
});
