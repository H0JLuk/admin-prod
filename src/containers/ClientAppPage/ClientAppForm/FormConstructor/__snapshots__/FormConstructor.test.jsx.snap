// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormConstructor /> test should be mount snap 1`] = `
"<Row className=\\"propertiesRow\\" gutter={{...}}>
  <Col className=\\"colFlex\\" span=\\"span\\">
    <FormItem rules=\\"rules\\" name=\\"name\\" className=\\"\\" label=\\"label\\" validateFirst={true}>
      <FormInputByType isEdit={false} value={{...}} />
    </FormItem>
  </Col>
</Row>"
`;

exports[`<FormConstructor /> test should render bannerType 1`] = `"<div class=\\"imageBlock\\"><div class=\\"imageContainer\\"><img class=\\"img\\" style=\\"background:linear-gradient(gradient)\\" src=\\"/images/clientAppDesign/vitrina_theme.png\\" alt=\\"banner\\"/></div></div>"`;

exports[`<FormConstructor /> test should render checkBoxes 1`] = `"<div class=\\"ant-checkbox-group\\"></div>"`;

exports[`<FormConstructor /> test should render div in default with editMode 1`] = `"<div class=\\"infoText\\">test</div>"`;

exports[`<FormConstructor /> test should render input in default 1`] = `"<span class=\\"ant-input-affix-wrapper\\"><input type=\\"text\\" value=\\"[object Object]\\" class=\\"ant-input\\"/><span class=\\"ant-input-suffix\\"><span role=\\"button\\" aria-label=\\"close-circle\\" tabindex=\\"-1\\" class=\\"anticon anticon-close-circle ant-input-clear-icon\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"close-circle\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z\\"></path></svg></span></span></span>"`;

exports[`<FormConstructor /> test should render textBlock 1`] = `"<textarea class=\\"ant-input\\">[object Object]</textarea>"`;
